Sources/JSONSchemaBuilder/JSONComponent/Modifier/Map.swift

```diff
@@ -0,0 +1,201 @@
+import JSONSchema
+
+extension JSONSchemaComponent {
+  public func map<NewOutput>(
+    _ transform: @Sendable @escaping (Output) -> NewOutput
+  ) -> JSONComponents.Map<Self, NewOutput> {
+    .init(upstream: self, transform: transform)
+  }
+
+  public func flatMap<NewComponent: JSONSchemaComponent>(
+    _ transform: @Sendable @escaping (Output) -> NewComponent
+  ) -> JSONComponents.FlatMap<NewComponent, Self> {
+    .init(upstream: self, transform: transform)
+  }
+
+  public func compactMap<NewOutput>(
+    _ transform: @Sendable @escaping (Output) -> NewOutput?
+  ) -> JSONComponents.CompactMap<Self, NewOutput> {
+    .init(upstream: self, transform: transform)
+  }
+}
+
+extension JSONComponents {
+  public struct Map<Upstream: JSONSchemaComponent, NewOutput>: JSONSchemaComponent {
+    public var definition: Schema {
+      upstream.definition
+    }
+
+    public var annotations: AnnotationOptions {
+      get {
+        upstream.annotations
+      }
+      set {
+        upstream.annotations = newValue
+      }
+    }
+
+    var upstream: Upstream
+    let transform: @Sendable (Upstream.Output) -> NewOutput
+
+    public init(upstream: Upstream, transform: @Sendable @escaping (Upstream.Output) -> NewOutput) {
+      self.upstream = upstream
+      self.transform = transform
+    }
+
+    public func validate(_ value: JSONValue) -> Validated<NewOutput, String> {
+      upstream.validate(value).map(transform)
+    }
+  }
+}
+
+extension JSONComponents {
+  public enum Conditional<First: JSONSchemaComponent, Second: JSONSchemaComponent>: JSONSchemaComponent
```

Comment:
Pull these out into separate files.

Use Conditional and Optionally in builders